### PR TITLE
Reconstruction for evicted objects

### PR DESCRIPTION
**Pull request by stephanie-wang**

### CODE REVIEW

File: src/common/redis_module/runtest.py [hunk 1]
```diff
@@ -131,7 +131,7 @@ def testObjectTableAddAndRemove(self):
     self.redis.execute_command("RAY.OBJECT_TABLE_REMOVE", "object_id1", "manager_id2")
     response = self.redis.execute_command("RAY.OBJECT_TABLE_LOOKUP", "object_id1")
     self.assertEqual(set(response), set())
-    # Remove a manager from an empty set, and make sure we still have an empty set.
+    # Remove a manager from an empty set, and make sure we now have an empty set.
```
Comment (pcmoritz) [inline src/common/redis_module/runtest.py]: "still" was right here I think

File: src/photon/photon_algorithm.c [hunk 2]
```diff
@@ -368,6 +368,14 @@ void handle_object_removed(local_scheduler_state *state, object_id object_id) {
   }
 }
 
+bool object_is_local(local_scheduler_state *state, object_id object_id) {
+  scheduling_algorithm_state *algorithm_state = state->algorithm_state;
+  available_object *entry;
+  HASH_FIND(handle, algorithm_state->local_objects, &object_id,
```
Comment (pcmoritz) [inline src/photon/photon_algorithm.c]: consider state->algorithm_state and getting rid of the first line

File: test/stress_tests.py [hunk 3]
```diff
@@ -108,5 +108,144 @@ def g(x):
...
+    num_args = 3
+    args = []
+    for i in range(num_args):
+      arg = no_dependency_task.remote(size)
+      args.append(arg)
+    for i in range(num_iterations):
+      args.append(multiple_dependency.remote(i, *args[i:i+num_args]))
+
+    # Get each value to force each task to finish. After some number of gets,
+    # old values should be evicted. Get each value again to force
```
Comment (pcmoritz) [inline test/stress_tests.py]: can you move the second sentence down where the getting again is done?

File: test/stress_tests.py [hunk 4]
```diff
@@ -108,5 +108,144 @@ def g(x):
...
+      return array
+
+    # Define a task with a single dependency, which returns its one argument.
+    @ray.remote
+    def single_dependency(i, arg):
+      arg = np.copy(arg)
+      arg[0] = i
+      return arg
+
+    # Launch num_iterations instances of the remote task, each dependent on the
```
Comment (pcmoritz) [inline test/stress_tests.py]: "each dependent on the one before it"

that isn't true, is it?

**Review by robertnishihara — COMMENTED** [review]
This looks good! I left some comments/questions to discuss.

File: src/common/state/redis.c [hunk 5]
```diff
@@ -465,24 +465,33 @@ void redis_object_table_lookup_callback(redisAsyncContext *c,
...
+    }
+    for (int j = 0; j < reply->elements; ++j) {
+      CHECK(reply->element[j]->type == REDIS_REPLY_STRING);
+      memcpy(managers[j].id, reply->element[j]->str, sizeof(managers[j].id));
+      redis_get_cached_db_client(db, managers[j], manager_vector + j);
+    }
+  } else {
+    LOG_FATAL("Unexpected reply type from object table lookup.");
   }
+
```
Comment (stephanie-wang) [inline src/common/state/redis.c]: Is that true? The original code looks like it would call the callback even if `manager_count = 0`.

File: lib/python/ray/worker.py [hunk 6]
```diff
@@ -446,8 +449,16 @@ def get_object(self, objectid):
     Args:
       objectid (object_id.ObjectID): The object ID of the value to retrieve.
     """
-    self.plasma_client.fetch([objectid.id()])
-    buff = self.plasma_client.get(objectid.id())
+    buff = self.plasma_client.get_local(objectid.id())
```
Comment (stephanie-wang) [inline lib/python/ray/worker.py]: I don't think reconstruction has to be serialized. We can call reconstruction on a bunch of objects serially, but the local scheduler's event loop should be able to process them asynchronously.

File: src/common/redis_module/runtest.py (hunk 1 above)
Comment (stephanie-wang) [inline src/common/redis_module/runtest.py]: I changed the API so that it returns `None` for no entry at all and an empty set for an empty entry.

File: src/photon/photon_scheduler.c [hunk 7]
```diff
@@ -162,8 +162,14 @@ void reconstruct_object_task_lookup_callback(object_id reconstruct_object_id,
...
                                              void *user_context) {
   /* Recursively resubmit the task and its task lineage to the scheduler. */
-  CHECKM(task != NULL,
-         "No task information found for object during reconstruction");
+  /* TODO(swang): The following check will fail if an object was created by a
+   * put. Once we're able to reconstruct those objects as well, uncomment the
+   * check and remove the if-check and exit. */
+  /* CHECKM(task != NULL,
+         "No task information found for object during reconstruction"); */
+  if (task == NULL) {
```
Comment (stephanie-wang) [inline src/photon/photon_scheduler.c]: We can reconstruct puts, but the rest of the code that's needed for that isn't in place yet. I was hoping to do those in another PR.

File: test/stress_tests.py (hunk 4 above)
Comment (stephanie-wang) [inline test/stress_tests.py]: Oops, yeah, I wrote the test case incorrectly.

File: python/ray/worker.py [hunk 8]
```diff
@@ -35,6 +35,10 @@
 LOG_SPAN_START = 1
 LOG_SPAN_END = 2
 
+# When performing ray.get, wait 1 second before attemping to reconstruct and
+# fetch the object again.
+GET_TIMEOUT = 1000
```
Comment (robertnishihara) [inline python/ray/worker.py]: rename to GET_TIMEOUT_MILLISECONDS

File: python/ray/worker.py [hunk 9]
```diff
@@ -439,18 +443,37 @@ def get_object(self, object_ids):
...
+      results = numbuf.retrieve_list(list(unready_ids.keys()),
+                                     self.plasma_client.conn,
+                                     GET_TIMEOUT)
+      # Remove any entries for objects we received during this iteration so we
+      # don't retrieve the same object twice.
+      for object_id, val in results:
+        if val is not None:
+          index = unready_ids[object_id]
+          final_results[index] = (object_id, val)
+          del unready_ids[object_id]
```
Comment (robertnishihara) [inline python/ray/worker.py]: Slightly cleaner I think to use `unready_ids.pop(object_id)`

File: python/ray/worker.py [hunk 10]
```diff
@@ -439,18 +443,37 @@ def get_object(self, object_ids):
...
+                       enumerate(final_results) if val is None)
+    # Try reconstructing any objects we haven't gotten yet. Try to get them
+    # until GET_TIMEOUT milliseconds passes, then repeat.
     while len(unready_ids) > 0:
-      results = numbuf.retrieve_list([object_id.id() for object_id in object_ids], self.plasma_client.conn, 1000)
-      unready_ids = [object_id for (object_id, val) in results if val is None]
-      # This would be a natural place to issue a command to reconstruct some of
-      # the objects.
+      for unready_id in unready_ids:
+        self.photon_client.reconstruct_object(unready_id)
```
Comment (robertnishihara) [inline python/ray/worker.py]: want to move the call to reconstruct to the end of the while loop so as to not flood the local scheduler with requests?

File: src/common/state/redis.c (hunk 5 above)
Comment (robertnishihara) [inline src/common/state/redis.c]: Oh I think you're right about that!

File: src/common/state/redis.c [hunk 11]
```diff
@@ -465,24 +465,33 @@ void redis_object_table_lookup_callback(redisAsyncContext *c,
...
-  }
-  for (int j = 0; j < reply->elements; ++j) {
-    CHECK(reply->element[j]->type == REDIS_REPLY_STRING);
-    memcpy(managers[j].id, reply->element[j]->str, sizeof(managers[j].id));
-    redis_get_cached_db_client(db, managers[j], manager_vector + j);
+
+  /* Parse the Redis reply. */
+  if (reply->type == REDIS_REPLY_NIL) {
+    /* The object entry did not exist. */
+    manager_count = -1;
```
Comment (robertnishihara) [inline src/common/state/redis.c]: Document in `object_table.h` that `-1` will be passed when there is no entry in the object table.

File: src/numbuf/CMakeLists.txt [hunk 12]
```diff
@@ -58,7 +59,8 @@ add_library(numbuf SHARED
   cpp/src/numbuf/sequence.cc
   python/src/pynumbuf/numbuf.cc
   python/src/pynumbuf/adapters/numpy.cc
-  python/src/pynumbuf/adapters/python.cc)
+  python/src/pynumbuf/adapters/python.cc
+  ${COMMON_EXTENSION})
```
Comment (robertnishihara) [inline src/numbuf/CMakeLists.txt]: This is to avoid code duplication for `PyStringToUniqueID`? Or some other reason?

File: src/photon/photon_scheduler.c (hunk 7 above)
Comment (robertnishihara) [inline src/photon/photon_scheduler.c]: This will still occur in some cases (e.g., when the put happened on the driver).

I'd suggest making it fatal so that we know when it happens (the alternative is deadlock, right?).

File: test/stress_tests.py [hunk 13]
```diff
@@ -108,5 +108,145 @@ def g(x):
...
+                     num_workers=1)
+
+  def tearDown(self):
+    # Clean up the Ray cluster.
+    self.assertTrue(ray.services.all_processes_alive())
+    ray.worker.cleanup()
+
+  def testSingleNodeSimple(self):
+    # Define the size of one task's return argument so that the combined sum of
+    # all objects' sizes is twice the plasma store's allotted memory.
```
Comment (robertnishihara) [inline test/stress_tests.py]: If you create an array of size `N`, then it will actually take up more than `N * *` bytes in the object store.

File: test/stress_tests.py [hunk 14]
```diff
@@ -108,5 +108,145 @@ def g(x):
...
+    num_objects = 1000
+    size = int(self.plasma_store_memory * 2 / (num_objects * 8))
```
Comment (robertnishihara) [inline test/stress_tests.py]: remove the `int(...)` and use `//` instead of `/`

File: test/stress_tests.py [hunk 15]
```diff
@@ -108,5 +108,145 @@ def g(x):
...
+
+    # Define a remote task with no dependencies, which returns a numpy array of
+    # the given size.
+    @ray.remote
+    def foo(i, size):
+      array = np.zeros(size)
+      array[0] = i
```
Comment (robertnishihara) [inline test/stress_tests.py]: probably unnecessary, but I'd be in favor of setting `array[-1] = i` as well to check that we're getting the whole object.

File: test/stress_tests.py (hunk 14 above)
Comment (robertnishihara) [inline test/stress_tests.py]: This also applies in `testSingleNodeRecursive`

Comment (robertnishihara) [inline test/stress_tests.py]: Since this is in `stress_tests.py`, can we make it `self.plasma_store_memory * 10` instead of `* 2`? (same with the other tests)

File: test/stress_tests.py [hunk 16]
```diff
@@ -108,5 +108,145 @@ def g(x):
...
+    # Get each value to force each task to finish. After some number of gets,
+    # old values should be evicted.
+    for i in range(num_iterations):
+      value = ray.get(args[i])
+      self.assertEqual(value[0], i)
+    # Get each value again to force reconstruction.
+    for i in range(num_iterations):
+      value = ray.get(args[i])
+      self.assertEqual(value[0], i)
+
```
Comment (robertnishihara) [inline test/stress_tests.py]: Can we either add another pass through in the reverse direction (in addition to what you have)? E.g.,

```python
for i in range(num_iterations)[::-1]:
  value = ray.get(args[i])
  self.assertEqual(value[0], i)
```

or randomly choose `100` values and reconstruction them, e.g.,

```python
for _ in range(100):
  i  = np.random.randint(num_objects)
  value = ray.get(args[i])
  self.assertEqual(value[0], i)
```

Actually, maybe add both?

File: test/stress_tests.py [hunk 17]
```diff
@@ -108,5 +108,145 @@ def g(x):
...
+      self.assertEqual(value[0], i)
+    # Get each value again to force reconstruction.
+    for i in range(num_objects):
+      value = ray.get(args[i])
+      self.assertEqual(value[0], i)
+
+  def testSingleNodeRecursive(self):
+    # Define the size of one task's return argument so that the combined sum of
+    # all objects' sizes is twice the plasma store's allotted memory.
+    num_iterations = 1000
```
Comment (robertnishihara) [inline test/stress_tests.py]: Can you call this `num_objects` to be consistent with the previous test?

Comment (robertnishihara) [inline test/stress_tests.py]: Same with the next test.

File: test/stress_tests.py [hunk 18]
```diff
@@ -108,5 +108,145 @@ def g(x):
...
+    # Launch num_args instances of the root task. Then launch num_iterations
+    # instances of the multi-dependency remote task, each dependent on the
+    # num_args tasks before it.
+    num_args = 3
+    args = []
+    for i in range(num_args):
+      arg = no_dependency_task.remote(size)
+      args.append(arg)
+    for i in range(num_iterations):
+      args.append(multiple_dependency.remote(i, *args[i:i+num_args]))
```
Comment (robertnishihara) [inline test/stress_tests.py]: `i+num_args` -> `(i + num_args)`

File: test/stress_tests.py (hunk 16 above)
Comment (robertnishihara) [inline test/stress_tests.py]: And do something similar in the next test?

File: python/plasma/plasma.py [hunk 19]
```diff
@@ -129,6 +129,10 @@ def create(self, object_id, size, metadata=None):
       metadata (buffer): An optional buffer encoding whatever metadata the user
         wishes to encode.
 
+    Returns:
+      A buffer for the object created. If the object wasn't able to be created
+        because it already exists locally, returns None.
+
     Raises:
       plasma_object_exists_error: This exception is raised if the object could
```
Comment (stephanie-wang) [inline python/plasma/plasma.py]: Oh, yes we're still using the exception. I think this comment is left over from when I was rebasing.

File: python/ray/worker.py (hunk 10 above)
Comment (stephanie-wang) [inline python/ray/worker.py]: I'm not sure how moving the call to the end of the loop will help?

File: src/numbuf/CMakeLists.txt (hunk 12 above)
Comment (stephanie-wang) [inline src/numbuf/CMakeLists.txt]: Yeah.

File: python/ray/worker.py (hunk 10 above)
Comment (robertnishihara) [inline python/ray/worker.py]: Well it will cause the worker to wait for GET_TIMEOUT before sending the reconstruct commands to the local scheduler, so if the get is fulfilled before GET_TIMEOUT passes, then no reconstruct commands will be sent. At least that's what I meant.

File: src/common/task.c [hunk 20]
```diff
@@ -182,6 +182,9 @@ task_spec *alloc_nil_task_spec(task_id task_id) {
 }
 
 int64_t task_spec_size(task_spec *spec) {
+  if (spec == NULL) {
```
Comment (pcmoritz) [inline src/common/task.c]: This should not be necessary, in which case is it called?

File: src/common/task.c [hunk 21]
```diff
@@ -315,7 +318,10 @@ struct task_impl {
 task *alloc_task(task_spec *spec,
                  scheduling_state state,
                  db_client_id local_scheduler_id) {
-  int64_t size = sizeof(task) - sizeof(task_spec) + task_spec_size(spec);
+  int64_t size = sizeof(task);
+  if (spec != NULL) {
```
Comment (pcmoritz) [inline src/common/task.c]: same as above

**Comment by robertnishihara** [comment]
One bug I encountered while testing this, which is not an issue with this PR and should be fixed separately (but is interesting nonetheless), is the following.

This is mostly an issue with the fact that functions are currently identified by their module + name, so functions with the same name override each other.

```python
import ray
import numpy as np

ray.init()

@ray.remote
def f(n):
  return np.zeros(n)

# Some of these objects are evicted.
l = [f.remote(100000000) for _ in range(1000)]

@ray.remote
def f(x, y):
  return x + y

# Do a tree reduction.
while len(l) > 1:
  l.append(f.remote(l.pop(0), l.pop(0)))

ray.get(l)
```

When some of the original tasks that created the list `l` are re-executed, the new definition of the function `f` will be used, and you will get an error like this

```
Remote function __main__.f failed with:

Traceback (most recent call last):
TypeError: f() missing 1 required positional argument: 'y'
```

File: src/common/task.c (hunk 20 above)
Comment (stephanie-wang) [inline src/common/task.c]: Oops, thanks for catching this! I was originally allocating an empty task using this, but turned out to not need it.

File: test/stress_tests.py [hunk 22]
```diff
@@ -108,5 +109,177 @@ def g(x):
...
+    # memory.
+    node_ip_address = "127.0.0.1"
+    self.redis_port = ray.services.new_port()
+    print(self.redis_port)
+    redis_address = ray.services.address(node_ip_address, self.redis_port)
+    self.plasma_store_memory = 10 ** 9
+    plasma_addresses = []
+    objstore_memory = (self.plasma_store_memory // self.num_local_schedulers)
+    for i in range(self.num_local_schedulers):
+        plasma_addresses.append(
```
Comment (robertnishihara) [inline test/stress_tests.py]: This line should be indented only 2 spaces

File: python/common/redis_module/runtest.py [hunk 23]
```diff
@@ -227,6 +222,38 @@ def testTaskTableAddAndLookup(self):
...
+    get_response = self.redis.execute_command("RAY.TASK_TABLE_GET", "task_id")
+    self.assertEqual(get_response, task_args[1:])
+
+    # If the current value is no longer the same as the test value, the
+    # response is nil.
+    response = self.redis.execute_command("RAY.TASK_TABLE_TEST_AND_UPDATE",
+                                          "task_id",
+                                          *task_args[:3])
+    self.assertEqual(response, None)
+    # Check that the update did not happen.
```
Comment (stephanie-wang) [inline python/common/redis_module/runtest.py]: Ah yes, you're right.

File: src/common/redis_module/ray_redis_module.c [hunk 24]
```diff
@@ -581,6 +586,35 @@ int ResultTableAdd_RedisCommand(RedisModuleCtx *ctx,
   return REDISMODULE_OK;
 }
 
+int ParseTaskState(RedisModuleString *state) {
```
Comment (stephanie-wang) [inline src/common/redis_module/ray_redis_module.c]: Thanks :)

File: src/common/redis_module/ray_redis_module.c [hunk 25]
```diff
@@ -821,6 +845,84 @@ int TaskTableUpdate_RedisCommand(RedisModuleCtx *ctx,
 }
 
 /**
+ * Update an entry in the task table. This does not update the task
```
Comment (stephanie-wang) [inline src/common/redis_module/ray_redis_module.c]: Ah yeah, dangers of copy pasta...

File: src/common/state/task_table.c [hunk 26]
```diff
@@ -30,6 +30,21 @@ void task_table_update(db_handle *db_handle,
...
 
+void task_table_test_and_update(db_handle *db_handle,
+                                OWNER task *task,
+                                scheduling_state test_state,
+                                retry_info *retry,
+                                task_table_get_callback done_callback,
+                                void *user_context) {
+  task_table_test_and_update_data *update_data =
+      malloc(sizeof(task_table_test_and_update_data));
+  update_data->task = task;
```
Comment (stephanie-wang) [inline src/common/state/task_table.c]: Yeah, this is a little unfortunate...I had to do it this way because of the way callback data is cleaned up (the failure callback calls `free(callback_data->data)`). Let me see how it would look if we kept the entire task in `task_table_test_and_update_data` instead of a pointer.

File: src/common/state/task_table.h [hunk 27]
```diff
@@ -87,6 +87,35 @@ void task_table_update(db_handle *db_handle,
...
+void task_table_test_and_update(db_handle *db_handle,
+                                OWNER task *task,
+                                scheduling_state test_state,
+                                retry_info *retry,
+                                task_table_get_callback done_callback,
+                                void *user_context);
+
+/* Data that is needed to test and set the task's scheduling state. */
+typedef struct {
+  task *task;
```
Comment (stephanie-wang) [inline src/common/state/task_table.h]: Oh yeah, I think the second is probably the best way.

File: src/common/state/task_table.c [hunk 28]
```diff
@@ -31,17 +31,21 @@ void task_table_update(db_handle *db_handle,
...
+                                task_id task_id,
                                 scheduling_state test_state,
+                                scheduling_state update_state,
                                 retry_info *retry,
                                 task_table_get_callback done_callback,
                                 void *user_context) {
   task_table_test_and_update_data *update_data =
       malloc(sizeof(task_table_test_and_update_data));
-  update_data->task = task;
+  memset(update_data, 0, sizeof(task_table_test_and_update_data));
```
Comment (robertnishihara) [inline src/common/state/task_table.c]: The memset is unnecessary, right?
